src/basic_recipes/barplot.jl

```diff
@@ -46,15 +49,60 @@ function AbstractPlotting.plot!(p::BarPlot)
         end
     end
 
-    bars = lift(p[1], p.fillto, p.width, in_y_direction) do xy, fillto, width, in_y_direction
+    bars = lift(p[1], p.fillto, p.width, p.dodge, p.x_gap, p.dodge_gap, p.stack, in_y_direction) do xy, fillto, width, dodge, x_gap, dodge_gap, stack, in_y_direction
+        
+        x = first.(xy)
+        y = last.(xy)
+        
         # compute half-width of bars
         if width === automatic
-            # times 0.8 for default gap
-            width = mean(diff(first.(xy))) * 0.8
-            width = ifelse(isfinite(width), width, 0.8)
+            x_unique = unique(filter(isfinite, x))
+            
+            if length(x_unique) == 1
+                width = 1.0
```

Comment:
A bit late for this comment, but I've just realized that this approach (setting default width to take the full space, and reducing it later) could be wrong. Now the `width` passed by the user is no longer respected, so for example in `barplot(1:3, rand(3), width=1)`, bars no longer touch (this is problematic for histograms, where there shouldn't be space between the bars).

The "proper way" IMO is to stick with the previous approach of defaulting to `width = 0.8 * default_gap`, and have the "dodge

Reply:
@greimel Is fixing this just a matter of changing defaults, where default `width` goes `0.8 * default gap` and default `x_gap` goes to `0`?

Reply:
You need to set both `x_gap` and `dodge_gap` to 0.

```julia
begin
	f = Figure(resolution = (800, 600))
	Axis(f[1, 1], xticks = 0.5:0.1:2.5)

	xs = 1:2
	ys = 0.5 .* sin.(xs)

	barplot!(xs, ys, color = :red, strokecolor = :black, width = 0.8, strokewidth = 0.1, x_gap = 0, dodge_gap = 0)

	f
end
```

<img width="651" alt="image" src="https://user-images.githubusercontent.com/6280307/115951750-c6cfbf80-a4e2-11eb-84a9-62f669c7e0d5.png">

We could set these two to `automatic` as well. `dodge_gap` would be zero if `dodge === automatic`. And `x_gap` would be zero if `width !== automatic`.

Reply:
Thanks for the explanation! I think the cleanest solution is to only use `x_gap` to determine the `width` if it is `automatic`.  This allows us to simplify those annoying `dodge` formulas as well, because there is no longer `x_gap` to worry about, see https://github.com/JuliaPlots/AbstractPlotting.jl/pull/702